base_layer/core/src/covenants/fields.rs

```diff
@@ -0,0 +1,329 @@
+        encoder::CovenentWriteExt,
+        error::CovenantError,
+    },
+    transactions::transaction::{TransactionInput, TransactionOutput},
+};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(u8)]
+pub enum OutputField {
+    Commitment = byte_codes::FIELD_COMMITMENT,
+    Script = byte_codes::FIELD_SCRIPT,
+    SenderOffsetPublicKey = byte_codes::FIELD_SENDER_OFFSET_PUBLIC_KEY,
+    Covenant = byte_codes::FIELD_COVENANT,
+    Features = byte_codes::FIELD_FEATURES,
+    FeaturesFlags = byte_codes::FIELD_FEATURES_FLAGS,
+    FeaturesMaturity = byte_codes::FIELD_FEATURES_MATURITY,
+    FeaturesUniqueId = byte_codes::FIELD_FEATURES_UNIQUE_ID,
+    FeaturesParentPublicKey = byte_codes::FIELD_FEATURES_PARENT_PUBLIC_KEY,
+    FeaturesMetadata = byte_codes::FIELD_FEATURES_METADATA,
+}
+
+impl OutputField {
+    pub fn from_byte(byte: u8) -> Result<Self, CovenantDecodeError> {
+        use byte_codes::*;
+        use OutputField::*;
+        match byte {
+            FIELD_COMMITMENT => Ok(Commitment),
+            FIELD_SCRIPT => Ok(Script),
+            FIELD_SENDER_OFFSET_PUBLIC_KEY => Ok(SenderOffsetPublicKey),
+            FIELD_COVENANT => Ok(Covenant),
+            FIELD_FEATURES => Ok(Features),
+            FIELD_FEATURES_FLAGS => Ok(FeaturesFlags),
+            FIELD_FEATURES_MATURITY => Ok(FeaturesMaturity),
+            FIELD_FEATURES_UNIQUE_ID => Ok(FeaturesUniqueId),
+            FIELD_FEATURES_PARENT_PUBLIC_KEY => Ok(FeaturesParentPublicKey),
+            FIELD_FEATURES_METADATA => Ok(FeaturesMetadata),
+
+            _ => Err(CovenantDecodeError::UnknownByteCode { code: byte }),
+        }
+    }
+
+    pub fn as_byte(&self) -> u8 {
+        *self as u8
+    }
+
+    pub fn get_field_value_ref<'a, T: 'static>(&self, output: &'a TransactionOutput) -> Option<&'a T> {
+        use OutputField::*;
+        let val = match self {
+            Commitment => &output.commitment as &dyn Any,
+            Script => &output.script as &dyn Any,
+            SenderOffsetPublicKey => &output.sender_offset_public_key as &dyn Any,
+            Covenant => unimplemented!(),
+            Features => &output.features as &dyn Any,
+            FeaturesFlags => &output.features.flags as &dyn Any,
+            FeaturesMaturity => &output.features.maturity as &dyn Any,
+            FeaturesUniqueId => &output.features.unique_id as &dyn Any,
+            FeaturesParentPublicKey => &output.features.parent_public_key as &dyn Any,
+            FeaturesMetadata => &output.features.metadata as &dyn Any,
+        };
+        val.downcast_ref::<T>()
+    }
+
+    pub fn get_field_value_bytes(&self, output: &TransactionOutput) -> Vec<u8> {
+        use OutputField::*;
+        match self {
+            Commitment => output.commitment.to_consensus_bytes(),
+            Script => output.script.to_consensus_bytes(),
+            SenderOffsetPublicKey => output.sender_offset_public_key.to_consensus_bytes(),
+            Covenant => unimplemented!(),
```

Comment:
Also to be implemented with the integration?

Reply:
Yeah covenant field is added in the integration and this code is updated

Reply:
Sweet